src/fairness/weighted_tree/test/load_accounting_db.cpp

```diff
@@ -0,0 +1,291 @@
+                                bank_name.c_str (),
+                                -1,
+                                NULL);
+        if (rc != SQLITE_OK) {
+            fprintf(stderr, "Failed to fetch data: %s\n", sqlite3_errmsg (DB));
+            sqlite3_close (DB);
+
+            return nullptr;
+        }
+
+        // execute SQL statement
+        rc = sqlite3_step (b_select_associations_stmt);
+        while (rc == SQLITE_ROW) {
+            std::string user = reinterpret_cast<char const *> (
+                sqlite3_column_text (b_select_associations_stmt, 0));
+            std::string single_user_shares = reinterpret_cast<char const *> (
+                sqlite3_column_text (b_select_associations_stmt, 1));
+            std::string job_usage = reinterpret_cast<char const *> (
+                sqlite3_column_text (b_select_associations_stmt, 3));
+            // place the user, their allocated shares, and
+            // their job_usage into vector
+            users.emplace_back (user, single_user_shares, job_usage);
+            rc = sqlite3_step (b_select_associations_stmt);
+        }
+        // iterate through associations and add them as children
+        // of node
+        for (unsigned int i = 0; i < users.size (); i++) {
+            std::string username = std::get<0> (users[i]);
+            std::string user_shares = std::get<1> (users[i]);
+            std::string user_usage = std::get<2> (users[i]);
+            user_node = std::make_shared<weighted_tree_node_t> (
+                            parent_bank,
+                            username,
+                            true,
+                            std::stoll (user_shares),
+                            std::stoll (user_usage));
+            node->add_child (user_node);
+            leaf_bank_usage += std::stoi (user_usage);
+        }
+        // aggregate usage from user nodes up to their respective banks
+        // and up to the root bank
+        for (std::shared_ptr<weighted_tree_node_t> curr_ancestor = node;
+             curr_ancestor != nullptr;
+             curr_ancestor = curr_ancestor->get_parent ()) {
+
+            curr_ancestor->set_usage (curr_ancestor->get_usage ()
+                                      + leaf_bank_usage);
+        }
+    }
+    // otherwise, this bank has sub banks, so call this
+    // function again with the first sub bank it found
+    else {
+        parent_bank = node;
+        while (rc == SQLITE_ROW) {
+            // execute SQL statement
+            std::string bank = reinterpret_cast<char const *> (
+                sqlite3_column_text (b_select_sub_banks_stmt, 0));
+            banks.push_back (bank);
+            rc = sqlite3_step (b_select_sub_banks_stmt);
+        }
+        for (std::string b : banks) {
+            // reset the prepared statements back to their initial state and
+            // clear their bindings
+            sqlite3_clear_bindings (b_select_associations_stmt);
```

Comment:
Same thing here. I would merge the `while` and `for` loops, so that all of the recursive calls and sqlite3_reset/clears are done within the while loop.  Then you can remove the `banks` vector.

Reply:
My apologies; maybe I could use some more clarification here.

My understanding of adding the sub banks to a vector and then iterating through that vector was so the recursive call would be executed for every sub bank found in the vector. Without it, would the function not exit after it found the first leaf bank?

Reply:
Oh. Good call.  We need to reset the prepared sql statements before the recursive call, but reseting within the loop would bork the `sqlite3_step`.  So as it is LGTM! My bad.